File: backend/evidence/rollcall/audits/parked_pool_audit.py
#!/usr/bin/env python3
"""Report the divided roll calls the campaign cannot import because their bill is unsigned.

Run from the repository root, against the extracted LegiScan datasets:

    python3 backend/evidence/rollcall/audits/parked_pool_audit.py

The campaign imports a roll call only when its measure became law. A bill that has
cleared both chambers but has not been signed sits at LegiScan status 3 ("enrolled"),
and its divided rolls wait. This sizes that pool per session and, more usefully, says
which part of it can ever move.

Three distinctions this makes, because they change what you should do:

  * A session that has adjourned (`sine_die` 1) will not sign anything more. Its parked
    rolls are dead weight, not a queue.
  * A concurrent or joint resolution is never presented to a governor, so its status can
    never reach enacted no matter how long you wait. Alaska's SCR 28/201/202 and
    California's SJR 7 are parked permanently.
  * A vetoed bill can keep status 3 in the dataset. Alaska HB 10 and HB 93 both read as
    "awaiting the governor" long after they were vetoed.

And one warning the Delaware run earned (CODE-FINDINGS §7 of legiscan-de-2163): a
non-zero live count here is NOT a reason to re-download the dataset. The status field
can trail the state's own record by more than a cut interval — two Delaware bills were
signed four days BEFORE the cut that still shows them unsigned. Check the state's own
action log instead, and note the rolls are already pending in `legislative_votes`, so
nothing needs re-fetching.
"""
from __future__ import annotations

import collections
import glob
import json
import os
import re

DATA = os.environ.get("LEGISCAN_DATA", "/Users/shu/legiscan-data")
CONFIG = "backend/src/pipeline/rollcall/legiscanStateConfigs.ts"

ENROLLED, ENACTED, VETOED = 3, 4, 5
# LegiScan bill_type codes for instruments a governor never signs.
UNPRESENTED_TYPES = {"CR", "JR", "R"}


def configured_sessions() -> set[str]:
    text = open(CONFIG).read()
    return {
        f"{jur}-{sid}"
        for jur, sid in re.findall(
            r'jurisdiction:\s*"([A-Z]{2})"[\s\S]{0,200}?sessionId:\s*(\d+)', text
        )
    }


def kept_question_patterns() -> dict[str, list[re.Pattern[str]]]:
    """jurisdiction -> the configured floor-question regexes."""
    text = open(CONFIG).read()
    out: dict[str, list[re.Pattern[str]]] = {}
    for jur, body in re.findall(r'jurisdiction:\s*"([A-Z]{2})"([\s\S]*?)\n  \},', text):
        pats = re.findall(r"pattern:\s*/\^?(.*?)\$?/[a-z]*\s*,\s*questionClass", body)
        out.setdefault(jur, [re.compile(f"^{p}$", re.I) for p in pats])
    return out


def is_divided(yea: int, nay: int) -> bool:
    low, high = min(yea, nay), max(yea, nay)
    return high > 0 and low >= high / 4.0


def main() -> None:
    configured = configured_sessions()
    kept = kept_question_patterns()
    live: list[tuple] = []
    adjourned: list[tuple] = []

    for path in sorted(glob.glob(f"{DATA}/??-[0-9]*")):
        name = os.path.basename(path)
        if not os.path.isdir(path) or not re.fullmatch(r"[a-z]{2}-\d+", name):
            continue
        jur, sid = name.split("-")
        jur = jur.upper()
        if f"{jur}-{sid}" not in configured:
            continue

        bills, sine = {}, None
        for p in glob.glob(f"{path}/*/*/bill/*.json"):
            try:
                bill = json.load(open(p))["bill"]
            except (OSError, ValueError):
                continue
            bills[bill["bill_id"]] = bill
            if sine is None:
                sine = bill.get("session", {}).get("sine_die")
        if not bills:
            continue

        patterns = kept.get(jur) or []
        rolls = collections.Counter()
        for p in glob.glob(f"{path}/*/*/vote/*.json"):
            try:
                rc = json.load(open(p))["roll_call"]
            except (OSError, ValueError):
                continue
            bill = bills.get(rc["bill_id"])
            if not bill or bill["status"] != ENROLLED:
                continue
            if not is_divided(rc["yea"], rc["nay"]):
                continue
            desc = (rc.get("desc") or "").strip()
            if patterns and not any(x.match(desc) for x in patterns):
                continue
            rolls[bill["bill_number"]] += 1
        if not rolls:
            continue

        never = sorted(b for b in rolls if bills_type(bills, b) in UNPRESENTED_TYPES)
        can_move = sorted(b for b in rolls if b not in never)
        row = (f"{jur}-{sid}", sum(rolls.values()), rolls, can_move, never)
        (live if sine == 0 else adjourned).append(row)

    report("STILL SITTING — these can still be signed", live, actionable=True)
    report("ADJOURNED — nothing more will be signed here", adjourned, actionable=False)


def bills_type(bills: dict, number: str) -> str:
    for bill in bills.values():
        if bill["bill_number"] == number:
            return bill.get("bill_type") or ""
    return ""


def report(title: str, rows: list[tuple], actionable: bool) -> None:
    print(f"\n{title}")
    if not rows:
        print("  (none)")
        return
    for session, total, rolls, can_move, never in sorted(rows, key=lambda r: -r[1]):
        movable = sum(rolls[b] for b in can_move)
        tail = (f"  — {movable} on bills a governor can still sign" if actionable
                else "  — the signing window has closed")
        print(f"  {session:10s} {total:3d} parked rolls / {len(rolls):2d} bills{tail}")
        if can_move and actionable:
            print(f"       check the state's action log for: {', '.join(can_move)}")
        elif can_move:
            print(f"       enrolled but never signed: {', '.join(can_move)}")
        if never:
            print(f"       never presented to a governor: {', '.join(never)}")
    print(f"  subtotal {sum(r[1] for r in rows)} rolls")
    if actionable:
        print("  Read the state's own action log for the bills above — do NOT assume a"
              "\n  re-download will show the signature. The rolls are already pending in"
              "\n  legislative_votes, so only the enactment fact is missing.")


if __name__ == "__main__":
    main()

File: backend/evidence/rollcall/audits/parkedPoolAudit.ts
/**
 * Report the divided floor roll calls the campaign cannot import yet because
 * LegiScan still shows their bill as enrolled (status 3) rather than enacted.
 *
 * Run from backend/, against the extracted LegiScan datasets:
 *
 *     npx tsx evidence/rollcall/audits/parkedPoolAudit.ts
 *
 * LEGISCAN_DATA overrides the dataset root (default /Users/shu/legiscan-data);
 * each configured session is read from `<root>/<st>-<sessionId>`.
 *
 * The filters are the production ones, not a reconstruction: every session in
 * LEGISCAN_STATE_CONFIGS supplies its own kept/excluded questions, and
 * classifyLegiscanRollCall decides floor-ness exactly as the fetcher does. A
 * roll counts as "divided" when the losing side holds at least a quarter of
 * the winning side — the campaign's worklist rule.
 *
 * The report splits the pool by whether the answer can still change:
 *
 *   * STILL SITTING (`sine_die` 0): the governor can still sign. Check the
 *     state's own action log — a dataset re-download is not the mechanism.
 *     LegiScan's status trails the state by more than a cut interval (two
 *     Delaware bills signed four days BEFORE the cut still read status 3;
 *     legiscan-de-2163 CODE-FINDINGS §7), and the rolls are already pending in
 *     legislative_votes, so only the enactment fact is missing.
 *   * ADJOURNED (`sine_die` 1): the disposition is settled, but status 3 does
 *     NOT mean "never signed" — it can simply be stale. Alabama 2021 SB46
 *     reads status 3 in the dataset and was signed 2021-05-17. Every bill here
 *     still needs the state's record read before it is written off.
 *
 * Status 3 proves nothing on its own in either direction: a vetoed bill keeps
 * it too (Alaska 2026 HB10 and HB93 read status 3 long after their vetoes).
 * Resolutions and constitutional amendments are flagged separately because a
 * governor's signature is usually not their enactment step.
 */
import { existsSync, readdirSync, readFileSync } from "node:fs";
import { join } from "node:path";
import {
  classifyLegiscanDatasetFile,
  classifyLegiscanRollCall,
  isLegiscanCommitteeChamberRollCall,
  parseLegiscanRollCall,
} from "../../../src/pipeline/rollcall/legiscanRollCall.js";
import { LEGISCAN_STATE_CONFIGS, type LegiscanStateConfig } from "../../../src/pipeline/rollcall/legiscanStateConfigs.js";

const DATA = process.env.LEGISCAN_DATA ?? "/Users/shu/legiscan-data";
const LEGISCAN_STATUS_ENROLLED = 3;

type SessionRow = {
  session: string;
  sineDie: number | null;
  rollsByBill: Map<string, number>;
  bills: string[];
  nonBills: string[];
};

function isDivided(yea: number, nay: number): boolean {
  const low = Math.min(yea, nay);
  const high = Math.max(yea, nay);
  return high > 0 && low >= high / 4;
}

function auditSession(config: LegiscanStateConfig, dir: string): SessionRow | null {
  const bills = new Map<number, { billNumber: string; billType: string; status: unknown }>();
  const votes: Record<string, unknown>[] = [];
  let sineDie: number | null = null;
  for (const entry of readdirSync(dir, { recursive: true }) as string[]) {
    if (!entry.endsWith(".json")) continue;
    let payload;
    try {
      payload = classifyLegiscanDatasetFile(JSON.parse(readFileSync(join(dir, entry), "utf8")) as unknown);
    } catch {
      continue;
    }
    if (payload.kind === "bill") {
      // Raw fields, not parseLegiscanBill: the audit only needs the status,
      // and the strict parser rejects numbers the fetcher never imports
      // (California's `AB160A`), which would abort the whole session.
      const { bill_id: billId, bill_number: billNumber, bill_type: billType, status, session } = payload.bill;
      if (typeof billId !== "number" || typeof billNumber !== "string" || typeof billType !== "string") continue;
      bills.set(billId, { billNumber, billType, status });
      const sine = (session as Record<string, unknown> | undefined)?.sine_die;
      if (sineDie === null && typeof sine === "number") sineDie = sine;
    } else if (payload.kind === "vote") {
      votes.push(payload.rollCall);
    }
  }
  if (bills.size === 0) return null;

  const rollsByBill = new Map<string, number>();
  const nonBills = new Set<string>();
  for (const raw of votes) {
    if (isLegiscanCommitteeChamberRollCall(raw)) continue;
    let rollCall;
    try {
      rollCall = parseLegiscanRollCall(raw);
    } catch {
      continue;
    }
    const bill = bills.get(rollCall.billId);
    if (!bill || bill.status !== LEGISCAN_STATUS_ENROLLED) continue;
    if (!isDivided(rollCall.yea, rollCall.nay)) continue;
    const verdict = classifyLegiscanRollCall({
      desc: rollCall.desc,
      total: rollCall.total,
      chamber: rollCall.chamber,
      billType: bill.billType,
      config,
      rollCallId: rollCall.rollCallId,
    });
    if (verdict.isFloorVote !== true) continue;
    rollsByBill.set(bill.billNumber, (rollsByBill.get(bill.billNumber) ?? 0) + 1);
    if (bill.billType !== "B") nonBills.add(bill.billNumber);
  }
  if (rollsByBill.size === 0) return null;
  const numbers = [...rollsByBill.keys()].sort();
  return {
    session: `${config.jurisdiction}-${config.sessionId}`,
    sineDie,
    rollsByBill,
    bills: numbers.filter((n) => !nonBills.has(n)),
    nonBills: numbers.filter((n) => nonBills.has(n)),
  };
}

function report(title: string, rows: SessionRow[], stillSitting: boolean): void {
  console.log(`\n${title}`);
  if (rows.length === 0) {
    console.log("  (none)");
    return;
  }
  const total = (row: SessionRow) => [...row.rollsByBill.values()].reduce((a, b) => a + b, 0);
  for (const row of rows.sort((a, b) => total(b) - total(a) || a.session.localeCompare(b.session))) {
    console.log(`  ${row.session.padEnd(10)} ${String(total(row)).padStart(3)} parked rolls / ${String(row.rollsByBill.size).padStart(2)} bills`);
    if (row.bills.length > 0) {
      const lead = stillSitting ? "can still be signed — read the state's action log for" : "dataset says enrolled — verify against the state's record";
      console.log(`       ${lead}: ${row.bills.join(", ")}`);
    }
    if (row.nonBills.length > 0) {
      console.log(`       not a bill (signature is usually not the enactment step): ${row.nonBills.join(", ")}`);
    }
  }
  console.log(`  subtotal ${rows.reduce((sum, row) => sum + total(row), 0)} rolls`);
  console.log(
    stillSitting
      ? "  Do NOT assume a re-download will show the signature; the rolls are already\n  pending in legislative_votes, only the enactment fact is missing."
      : "  Status 3 after adjournment can be stale (Alabama 2021 SB46 was signed\n  2021-05-17 and still reads 3). Verify before treating any of these as dead.",
  );
}

function main(): void {
  const seen = new Set<string>();
  const sitting: SessionRow[] = [];
  const adjourned: SessionRow[] = [];
  for (const config of Object.values(LEGISCAN_STATE_CONFIGS)) {
    const key = `${config.jurisdiction}-${config.sessionId}`;
    if (seen.has(key)) continue;
    seen.add(key);
    const dir = join(DATA, `${config.jurisdiction.toLowerCase()}-${config.sessionId}`);
    if (!existsSync(dir)) continue;
    const row = auditSession(config, dir);
    if (row) (row.sineDie === 0 ? sitting : adjourned).push(row);
  }
  report("STILL SITTING — these can still be signed", sitting, true);
  report("ADJOURNED — disposition settled, dataset status unverified", adjourned, false);
}

main();
